### DISCUSSION TITLE
Reading CSV output

### DISCUSSION THREAD

**Discussion by megasuperlexa** [DarkWanderer/ClickHouse.Client#442] [Q&A]
Hello, I was hoping to use Clickhouse csv output capability but unsure how to do it using Clickhouse.Client.
The following code:

```csharp
            var cmd = conn.CreateCommand();
            cmd.CommandText = "select * from my_distributed_table FORMAT CSVWithNames";
            conn.Open();

            using var reader = await cmd.ExecuteReaderAsync(ct);
            using var writer = new StreamWriter("file.csv");

            while (!ct.IsCancellationRequested && await reader.ReadAsync(ct))
            {
                writer.WriteLine(reader.GetValue(0).ToString());
            }
```

throws an exception on `await cmd.ExecuteReaderAsync(ct)`: "Attempted to read past the end of the stream."
It is of course works without `FORMAT CSVWithNames` but my point is to read in csv format.
Any ideas?

**Comment by megasuperlexa** [comment]
looking into implementation, the best way to do it seem to be
```csharp
using var result = await cmd.ExecuteRawResultAsync(ct);
await using var chStream = await result.ReadAsStreamAsync().WaitAsync(ct);
...
await chStream.CopyToAsync(writer, ct);
```

**Comment by DarkWanderer** [comment]
Hi. Indeed, this function was created for this very purpose - glad you found it useful
